File: sdk/typescript/examples/02-staking.ts
/**
 * Example 2: Staking and Delegation
 * 
 * Demonstrates:
 * - Viewing validators
 * - Delegating stake
 * - Checking rewards
 * - Claiming rewards
 */

import { AetherClient, AetherKeypair, StakingHelper } from '@aether/sdk';

async function main() {
  const client = new AetherClient({ rpcUrl: 'http://localhost:8545' });
  const keypair = await AetherKeypair.fromSeed('my seed phrase');

  const staking = new StakingHelper(client, keypair);

  console.log('Fetching validators...');
  const validators = await staking.getValidators();
  
  console.log(`Found ${validators.length} active validators`);
  for (const v of validators.slice(0, 5)) {
    console.log(`- ${v.address}: ${v.stake} stake, ${v.commission/100}% commission`);
  }

  const bestValidator = validators.sort((a, b) => 
    (b.uptime - a.uptime) || (a.commission - b.commission)
  )[0];

  console.log(`\nDelegating to best validator: ${bestValidator.address}`);
  
  const delegateTx = await staking.delegate(bestValidator.address, 10000n);
  const txHash = await client.sendTransaction(delegateTx);
  console.log('Delegation tx:', txHash);
  
  await client.waitForTransaction(txHash);
  console.log('Delegation confirmed');

  console.log('\nChecking pending rewards...');
  const rewards = await staking.getPendingRewards(keypair.address);
  console.log('Pending rewards:', rewards.toString(), 'AIC');

  if (rewards > 0n) {
    console.log('Claiming rewards...');
    const claimTx = await staking.claimRewards();
    const claimHash = await client.sendTransaction(claimTx);
    await client.waitForTransaction(claimHash);
    console.log('Rewards claimed!');
  }

  const delegation = await staking.getDelegation(keypair.address, bestValidator.address);
  console.log('\nDelegation info:');
  console.log('- Amount:', delegation?.amount.toString());
  console.log('- Rewards:', delegation?.rewards.toString());
}

main().catch(console.error);

File: sdk/python/examples/02_staking.py
"""
Example 2: Staking and Delegation

Demonstrates:
- Viewing validators
- Delegating stake
- Checking rewards
- Claiming rewards
"""

import asyncio
from aether import AetherClient, Keypair, StakingHelper


async def main():
    async with AetherClient(rpc_url="http://localhost:8545") as client:
        keypair = Keypair.from_seed("my seed phrase")

        staking = StakingHelper(client, keypair)

        print("Fetching validators...")
        validators = await staking.get_validators()

        print(f"Found {len(validators)} active validators")
        for v in validators[:5]:
            print(f"- {v.address}: {v.stake} stake, {v.commission/100}% commission")

        best_validator = max(
            validators,
            key=lambda v: (v.uptime, -v.commission)
        )

        print(f"\nDelegating to best validator: {best_validator.address}")

        delegate_tx = await staking.delegate(best_validator.address, 10000)
        tx_hash = await client.send_transaction(delegate_tx)
        print(f"Delegation tx: {tx_hash}")

        await client.wait_for_transaction(tx_hash)
        print("Delegation confirmed")

        print("\nChecking pending rewards...")
        rewards = await staking.get_pending_rewards(keypair.address)
        print(f"Pending rewards: {rewards} AIC")

        if rewards > 0:
            print("Claiming rewards...")
            claim_tx = await staking.claim_rewards()
            claim_hash = await client.send_transaction(claim_tx)
            await client.wait_for_transaction(claim_hash)
            print("Rewards claimed!")

        delegation = await staking.get_delegation(keypair.address, best_validator.address)
        if delegation:
            print("\nDelegation info:")
            print(f"- Amount: {delegation.amount}")
            print(f"- Rewards: {delegation.rewards}")


if __name__ == "__main__":
    asyncio.run(main())

